### PR TITLE
update for stop loss

### LINKED ISSUE #170: Update Cosmwasm LOB contract to work with new Javascript backend
### Background
Backend script was changed from Python to Javascript to add stop-loss feature here https://github.com/VolumeFi/paloma/issues/156

### Done when

- [x] Cosmwasm contract https://github.com/VolumeFi/limit-order-bot-uni-v2-cosmwasm is compatible with new Javascript backend script

### PR DESCRIPTION
**Pull request by webelf101**
fixes https://github.com/VolumeFi/paloma/issues/170

### CODE REVIEW

File: src/contract.rs [hunk 1]
```diff
@@ -55,20 +57,31 @@ pub mod execute {
         deps: DepsMut,
         env: Env,
         deposit_ids: Vec<u32>,
+        profit_taking_or_stop_loss: Vec<bool>,
```
Comment (measure-fi) [inline src/contract.rs]: As written here (and most other places) I cannot tell what this `bool` is meant to represent (what does `true` mean? That we're both profit taking and stop loss? That we're one or the other. Reading just this line I think I would assume `true` means `stop_loss` but later it turns out that is not the case). Anyway the normal way to do this in rust is via an enum. So something like

```
enum Reason {
    ProfitTaking,
    StopLoss,
}
```

Then this becomes a `Vec<Reason>` and it just overall makes more sense. But this is part of the public API (I think?) so maybe you just want a `profit_taking: bool`.

File: src/contract.rs [hunk 2]
```diff
@@ -55,20 +57,31 @@ pub mod execute {
...
     ) -> Result<Response<PalomaMsg>, ContractError> {
         let state = STATE.load(deps.storage)?;
+        if deposit_ids.len() != profit_taking_or_stop_loss.len() {
```
Comment (measure-fi) [inline src/contract.rs]: This condition winds up being a little wonky, so what I think you'd like is:

```
#[cw_serde]
pub struct Deposit {
        id: u32,
        profit_taking: bool,
}

#[cw_serde]
pub enum ExecuteMsg {
    PutWithdraw(Vec<Deposit>),
}
```

File: src/contract.rs [hunk 3]
```diff
@@ -80,27 +93,35 @@ pub mod execute {
             fallback: false,
         };
 
-        let mut tokens: Vec<Token> = vec![];
+        let mut tokens_id: Vec<Token> = vec![];
+        let mut tokens_take_profit: Vec<Token> = vec![];
         let retry_delay: u64 = RETRY_DELAY.load(deps.storage)?;
-        for deposit_id in deposit_ids {
+        let mut i = 0;
+        while i < deposit_ids.len() {
```
Comment (measure-fi) [inline src/contract.rs]: This is not the way to do a for loop. Use
```
for deposit in deposits
```
or similar.

Comment (measure-fi) [reply src/contract.rs]: (
for this _exact_ circumstance you would normally use one of the zip functions
https://doc.rust-lang.org/std/iter/fn.zip.html
or
https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.zip
)

**Comment by taariq** [comment]
@webelf101

File: src/contract.rs (hunk 2 above)
Comment (webelf101) [reply src/contract.rs]: fixed

File: src/contract.rs (hunk 3 above)
Comment (webelf101) [reply src/contract.rs]: fixed

File: src/contract.rs (hunk 1 above)
Comment (webelf101) [reply src/contract.rs]: I'd like to leave it as bool type to match with Vyper contract type.

Comment (measure-fi) [reply src/contract.rs]: Fair. I still would prefer it were renamed.

File: src/contract.rs [hunk 4]
```diff
@@ -80,27 +87,33 @@ pub mod execute {
...
-        let mut tokens: Vec<Token> = vec![];
+        let mut tokens_id: Vec<Token> = vec![];
+        let mut tokens_take_profit: Vec<Token> = vec![];
         let retry_delay: u64 = RETRY_DELAY.load(deps.storage)?;
-        for deposit_id in deposit_ids {
+        for deposit in deposits {
+            let deposit_id = deposit.deposit_id;
+            let profit_taking = deposit.profit_taking_or_stop_loss;
             if let Some(timestamp) = WITHDRAW_TIMESTAMP.may_load(deps.storage, deposit_id)? {
                 if timestamp.plus_seconds(retry_delay).lt(&env.block.time) {
```
Comment (measure-fi) [inline src/contract.rs:97]: This is the same bug as indicated in the audit PR's, but it's not part of this PR I guess.

**Review by measure-fi — APPROVED** [review]

File: src/contract.rs (hunk 4 above)
Comment (webelf101) [reply src/contract.rs:97]: Bug? What's the bug exactly? @measure-fi

Comment (measure-fi) [reply src/contract.rs:97]:
```
if let Some(timestamp) = WITHDRAW_TIMESTAMP.may_load(deps.storage, deposit_id)? {
    if timestamp.plus_seconds(retry_delay).lt(&env.block.time) {
        ...
    } else {
        // Should there be code to handle this case?
        // Also as mentioned, I expect it's the same as in the else block below;
        // there are a couple ways to make it so you don't have to copy/paste it.
    }
} else {
    ...
}
```
